src/passes/OptimizeInstructions.cpp

```diff
@@ -1804,6 +1823,41 @@ struct OptimizeInstructions
       // Don't bother when `ifFalse` isn't pure - we would need to reverse the
       // order using a temp local, which would be bad
     }
+    {
+      // TODO: Remove this after landing SCCP pass. See: #4161
+
+      // i32(x) ? i32(x) : 0  ==>  x
+      Expression *x, *y;
+      if (matches(curr, select(any(&x), i32(0), any(&y))) &&
+          areConsecutiveInputsEqualAndFoldable(x, y)) {
+        return curr->ifTrue;
+      }
+      // i32(x) ? 0 : i32(x)  ==>  0
+      if (matches(curr, select(i32(0), any(&x), any(&y))) &&
+          areConsecutiveInputsEqualAndRemovable(x, y)) {
+        return curr->ifTrue;
```

Comment:
This will not optimize the case where they are foldable but not removable, which is exactly the  case of a tee and a get.

To handle that case, this code path could check for them being foldable, and then not remove x, but instead return `(x, 0)` - that is, a block with a drop of x, and then the constant 0. (Other passes will remove the dropped x and the block later, if that is possible.)